File: api/spring-boot-api/src/main/java/com/spicytodo/controller/TodoController.java
package com.spicytodo.controller;

import com.spicytodo.dto.TodoCreateRequest;
import com.spicytodo.dto.TodoUpdateRequest;
import com.spicytodo.model.Todo;
import com.spicytodo.model.TodoStats;
import com.spicytodo.service.TodoService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api")
@CrossOrigin(origins = {"http://localhost:3000", "http://127.0.0.1:3000"})
public class TodoController {

    @Autowired
    private TodoService todoService;

    @GetMapping("/todos")
    public ResponseEntity<List<Todo>> getAllTodos(
            @RequestParam(required = false) String filter,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String priority) {
        List<Todo> todos = todoService.getAllTodos(filter, search, priority);
        return ResponseEntity.ok(todos);
    }

    @GetMapping("/todos/{id}")
    public ResponseEntity<Todo> getTodoById(@PathVariable String id) {
        return todoService.getTodoById(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @PostMapping("/todos")
    public ResponseEntity<Todo> createTodo(@Valid @RequestBody TodoCreateRequest request) {
        Todo todo = todoService.createTodo(
                request.getText(),
                request.getPriority(),
                request.getCompleted(),
                request.getDueDate(),
                request.getReminderTime()
        );
        return ResponseEntity.status(HttpStatus.CREATED).body(todo);
    }

    @PutMapping("/todos/{id}")
    public ResponseEntity<Todo> updateTodo(@PathVariable String id, @Valid @RequestBody TodoUpdateRequest request) {
        return todoService.updateTodo(
                id,
                request.getText(),
                request.getPriority(),
                request.getCompleted(),
                request.getDueDate(),
                request.getReminderTime()
        ).map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @DeleteMapping("/todos/{id}")
    public ResponseEntity<Map<String, String>> deleteTodo(@PathVariable String id) {
        if (todoService.deleteTodo(id)) {
            return ResponseEntity.ok(Map.of("message", "Todo deleted successfully"));
        }
        return ResponseEntity.notFound().build();
    }

    @PatchMapping("/todos/{id}/toggle")
    public ResponseEntity<Todo> toggleTodo(@PathVariable String id) {
        return todoService.toggleTodo(id)
                .map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping("/todos/stats/summary")
    public ResponseEntity<TodoStats> getStats() {
        return ResponseEntity.ok(todoService.getStats());
    }

    @DeleteMapping("/todos/completed")
    public ResponseEntity<Map<String, String>> clearCompleted() {
        todoService.clearCompleted();
        return ResponseEntity.ok(Map.of("message", "Completed todos cleared"));
    }
}

File: api/dotnet-api/Controllers/TodoController.cs
using Microsoft.AspNetCore.Mvc;
using DotnetApi.Models;
using DotnetApi.Services;

namespace DotnetApi.Controllers
{
    [ApiController]
    [Route("api/todos")]
    public class TodoController : ControllerBase
    {
        private readonly TodoService _todoService;

        public TodoController(TodoService todoService)
        {
            _todoService = todoService;
        }

        [HttpGet]
        public ActionResult<IEnumerable<Todo>> GetAllTodos([FromQuery] string? filter, [FromQuery] string? search, [FromQuery] string? priority)
        {
            var todos = _todoService.GetAllTodos(filter, search, priority);
            return Ok(todos);
        }

        [HttpGet("{id}")]
        public ActionResult<Todo> GetTodoById(string id)
        {
            var todo = _todoService.GetTodoById(id);
            if (todo == null)
                return NotFound(new { error = "Todo not found" });

            return Ok(todo);
        }

        [HttpPost]
        public ActionResult<Todo> CreateTodo([FromBody] TodoCreateRequest request)
        {
            var todo = _todoService.CreateTodo(
                request.Text,
                request.Priority,
                request.Completed,
                request.DueDate,
                request.ReminderTime
            );

            return CreatedAtAction(nameof(GetTodoById), new { id = todo.Id }, todo);
        }

        [HttpPut("{id}")]
        public ActionResult<Todo> UpdateTodo(string id, [FromBody] TodoUpdateRequest request)
        {
            var todo = _todoService.UpdateTodo(
                id,
                request.Text,
                request.Priority,
                request.Completed,
                request.DueDate,
                request.ReminderTime
            );

            if (todo == null)
                return NotFound(new { error = "Todo not found" });

            return Ok(todo);
        }

        [HttpDelete("{id}")]
        public ActionResult DeleteTodo(string id)
        {
            if (!_todoService.DeleteTodo(id))
                return NotFound(new { error = "Todo not found" });

            return Ok(new { message = "Todo deleted successfully" });
        }

        [HttpPatch("{id}/toggle")]
        public ActionResult<Todo> ToggleTodo(string id)
        {
            var todo = _todoService.ToggleTodo(id);
            if (todo == null)
                return NotFound(new { error = "Todo not found" });

            return Ok(todo);
        }

        [HttpGet("stats/summary")]
        public ActionResult<TodoStats> GetStats()
        {
            var stats = _todoService.GetStats();
            return Ok(stats);
        }

        [HttpDelete("completed")]
        public ActionResult ClearCompleted()
        {
            _todoService.ClearCompleted();
            return Ok(new { message = "Completed todos cleared" });
        }
    }
}

